integrations/Criteo/browser.js

```diff
@@ -0,0 +1,222 @@
+/* eslint-disable no-unused-expressions */
+import md5 from "md5";
+import logger from "../../utils/logUtil";
+import ScriptLoader from "../ScriptLoader";
+
+class Criteo {
+  constructor(config) {
+    this.name = "Criteo";
+    this.hashMethod = config.hashMethod;
+    this.accountId = config.accountId;
+    this.url = config.homePageUrl;
+    // eslint-disable-next-line no-nested-ternary
+    this.deviceType = /iPad/.test(navigator.userAgent)
+      ? "t"
+      : /Mobile|iP(hone|od)|Android|BlackBerry|IEMobile|Silk/.test(
+          navigator.userAgent
+        )
+      ? "m"
+      : "d";
+    this.fieldMapping = config.fieldMapping;
+    this.OPERATOR_LIST = ["eq", "gt", "lt", "ge", "le", "in"];
+  }
+
+  init() {
+    logger.debug("===in init Criteo===");
+    if (!this.accountId) {
+      return;
+    }
+    window.criteo_q = window.criteo_q || [];
+
+    ScriptLoader(
+      "Criteo",
+      `//dynamic.criteo.com/js/ld/ld.js?a=${this.accountId}`
+    );
+    window.criteo_q.push({ event: "setAccount", account: this.accountId });
+    window.criteo_q.push({ event: "setSiteType", type: this.deviceType });
+  }
+
+  // eslint-disable-next-line class-methods-use-this
+  isLoaded() {
+    logger.debug("in Criteo isLoaded");
+    return !!(window.criteo_q && window.criteo_q.push !== Array.prototype.push);
+  }
+
+  // eslint-disable-next-line class-methods-use-this
+  isReady() {
+    logger.debug("in Criteo isReady");
+    return !!(window.criteo_q && window.criteo_q.push !== Array.prototype.push);
+  }
+
+  handleCommonFields(rudderElement) {
+    const { message } = rudderElement;
+    const { properties } = message;
+
+    const setEmail = {};
+    const setZipcode = {};
+
+    const finalRequest = [
+      { event: "setCustomerId", id: md5(message.userId) },
+      { event: "setRetailerVisitorId", id: md5(message.anonymousId) },
+    ];
+
+    if (properties.email) {
+      const email = properties.email.trim().toLowerCase();
+      setEmail.event = "setEmail";
+      setEmail.hash_method = this.hashMethod;
+      setEmail.email = this.hashMethod === "md5" ? md5(email) : email;
+      finalRequest.push(setEmail);
+    }
+
+    if (properties.zipCode) {
+      setZipcode.event = "setZipcode";
+      setZipcode.zipCode = properties.zipCode || properties.zip;
+      finalRequest.push(setZipcode);
+    }
+
+    return finalRequest;
+  }
+
+  page(rudderElement) {
+    const { name, properties } = rudderElement.message;
+
+    const finalPayload = this.handleCommonFields(rudderElement);
+
+    if (
+      name === "home" ||
+      (this.url && this.url === window.location.href) ||
+      properties.url === this.url
+    ) {
+      const homeEvent = {
+        event: "viewHome",
+      };
+      finalPayload.push(homeEvent);
+    }
+
+    window.criteo_q.push(finalPayload);
+  }
+
+  track(rudderElement) {
+    const { event, properties } = rudderElement.message;
```

Comment:
we should add a check if `properties` is empty or not present.

Reply:
handled